bin/release.php

```diff
@@ -0,0 +1,158 @@
+		basename( __FILE__ ) . ' $milestone_from $milestone_to'
+	);
+}
+$milestone_from = $argv[1];
+$milestone_to   = $argv[2];
+
+// Update and commit version numbers.
+echo "Updating and committing version numbers...\n";
+shell_exec( 'bin/update-version-numbers.sh ' . $milestone_to );
+echo "DONE\n";
+
+// Update and commit the changelog.
+echo "Updating and committing the changelog...\n";
+$release_log = generate_release_log( $milestone_to );
+write_to_changelog_file( $release_log, $milestone_from, $milestone_to );
+shell_exec(
+	'git add CHANGELOG.md && git commit -m "Update CHANGELOG.md for version "' .
+	$milestone_to
+);
+echo "DONE! Please review the changes and amend if needed\n";
+
+// Show a prompt to create a PR on GitHub.
+echo 'Do you want to create a release PR on GitHub? [y/n]';
+$confirmation = trim( fgets( STDIN ) );
+if ( $confirmation === 'y' ) {
+	create_pull_request( $milestone_to, $release_log );
+}
+
+/**
+ * Generates the changelog for the passed milestone.
+ *
+ * @param string $milestone The milestone for which to generate the changelog.
+ * @return string The resulting changelog for the specified milestone.
+ */
+function generate_release_log( string $milestone ): string {
+	$result = '';
+	$delay  = 500; // Delay between API calls in milliseconds.
+	$labels = array( // GitHub labels for which to get data.
+		'Added',
+		'Changed',
+		'Deprecated',
+		'Removed',
+		'Fixed',
+		'Security',
+	);
+
+	foreach ( $labels as $label ) {
+		echo "Getting PRs with label \"{$label}\"...\n";
+		$data = get_pull_request_data( $milestone, 'Changelog: ' . $label );
+		if ( ! empty( $data ) ) {
+			$result .= "### {$label}\n\n{$data}\n";
+		}
+
+		usleep( $delay * 1000 );
+	}
+
+	// Include a dependency updates link if there are any dependency updates.
+	echo "Adding \"Dependency Updates\" entry\n";
+	$data = get_pull_request_data( $milestone, 'Component: Dependencies', 1 );
+	if ( ! empty( $data ) ) {
+		$link    = "- The list of all dependency updates for this release is available [here](https://github.com/Parsely/wp-parsely/pulls?q=is%3Apr+is%3Amerged+milestone%3A{$milestone}+label%3A%22Component%3A+Dependencies%22).";
+		$result .= "### Dependency Updates\n\n{$link}\n";
+	}
+
+	return $result;
+}
+
+/**
+ * Gets the data of the Pull Requests that comply to the passed arguments.
+ *
+ * @param string $milestone The milestone of the PRs that we want to fetch.
+ * @param string $label The label of the PRs that we want to fetch.
+ * @param int    $limit The maximum number of results to be returned.
+ * @return string|false|null The result of the `shell_exec` function.
+ */
+function get_pull_request_data( string $milestone, string $label, $limit = 100 ) {
+	return shell_exec(
+		'gh pr list --repo Parsely/wp-parsely --search "milestone:' . $milestone .
+		'" --label "' . $label . '"  --state merged --limit ' . $limit .
+		' --json title,number,url --jq \'.[] | "- \(.title) ([#\(.number)](\(.url)))"\''
```

Comment:
Although this is fine but I think having each argument on new line is more easy to read (same feedback below where we are creating PR)

Reply:
I agree, but do not want to touch this currently 🙂